packages/server/src/core/internals/procedureBuilder.ts

```diff
@@ -31,19 +31,35 @@ type CreateProcedureReturnInput<
   _output_out: FallbackValue<TNext['_output_out'], TPrev['_output_out']>;
 }>;
 
+type OverwriteIfDefined<T, K> = UnsetMarker extends T ? K : FlatOverwrite<T, K>;
+
+type ErrorMessage<T extends string> = T;
+
 export interface ProcedureBuilder<TParams extends ProcedureParams> {
   /**
    * Add an input parser to the procedure.
    */
   input<$TParser extends Parser>(
-    schema: $TParser,
+    schema: TParams['_input_out'] extends UnsetMarker
+      ? $TParser
+      : inferParser<$TParser>['out'] extends Record<string, unknown>
+      ? TParams['_input_out'] extends Record<string, unknown>
+        ? $TParser
+        : ErrorMessage<'Both input parsers did not resolve to an object'>
+      : ErrorMessage<'Both input parsers did not resolve to an object'>,
```

Comment:
Good catch @juliusmarminge 

```suggestion
        : ErrorMessage<'All input parsers did not resolve to an object'>
      : ErrorMessage<'All input parsers did not resolve to an object'>,
```